File: LeetCode_Algorithm/[re]1192(Critical Connections in a Network)/1192.py
from typing import List
class Solution:
    def criticalConnections(self, n: int, connections: List[List[int]]) -> List[List[int]]:
        # sol1 3694ms 52%
        # graph를 그려놓고 dfs를 통해 그래프 안의 루프를 찾는 방법
        # tarjan algorithm
        from collections import defaultdict
        self.hashMap = defaultdict(set)
        for [i,j] in connections:
            self.hashMap[i].add(j)
            self.hashMap[j].add(i)
        self.rank = [-1] * n; self.ans = []
        # depth와 현재 노드를 넣고 dfs
        self.dfs(None, 0, 0)
        return self.ans
    
    def dfs(self, prevNode, currNode, currDepth):
        self.rank[currNode] = currDepth
        minDepth = currDepth
        for node in self.hashMap[currNode]:
            if node == prevNode:
                continue
            temp = self.rank[node]
            # 해당 노드에 도달하지 못했을 경우 dfs
            if temp == -1:
                temp = self.dfs(currNode, node, currDepth+1)
            # 현재 depth보다 들어간 depth가 더 크다면 그것은 critical node
            if temp > currDepth:
                self.ans.append([currNode, node])
            else:
                minDepth = min(minDepth, temp)
        self.rank[currNode] = minDepth
        return minDepth   
print(Solution().criticalConnections(4, [[0,1],[1,2],[2,0],[1,3]]))

File: LeetCode_Algorithm/[re]1192(Critical Connections in a Network)/1192.js
var dfs = function(prevNode, currNode, currDepth) {
  rank[currNode] = currDepth
  minDepth = currDepth
  for (let node of hashMap[currNode]) {
    if (prevNode == node) {
      continue
    }
    temp = rank[node]
    if (temp == -1) {
      temp = dfs(currNode, node, currDepth+1)
    }
    console.log(prevNode, currNode, node, temp)
    if (temp > currDepth) {
      ans.push([currNode, node])
    } else {
      minDepth = Math.min(minDepth, temp)
    }
  }
  rank[currNode] = minDepth
  return rank[currNode]
}
var criticalConnections = function(n, connections) {
  // sol1
  hashMap = new Map()
  for (let i=0; i < n; i++) {
    hashMap[i] = new Set()
  }
  for (let [i,j] of connections) {
    hashMap[i].add(j)
    hashMap[j].add(i)
  }
  rank = new Array(n).fill(-1)
  ans = []
  dfs(null, 0, 0)
  console.log(rank)
  return ans
};

console.log(criticalConnections(4, [[0,1],[1,2],[2,0],[1,3]]))